src/lib/CofEnv.ml

```diff
@@ -4,55 +4,99 @@ module D = Domain
 module UF = DisjointSet.Make (PersistentTable.M)
 module VarSet = Set.Make (Int)
 
-type env =
+type reduced_env =
   { classes : D.dim UF.t;
-    (** equivalence classes of dimensions *)
+    (** equivalence classes of dimensions from reduced cofibrations *)
 
-    cof : D.cof;
-    (** a cofibration which is assumed true, but has not been eliminated yet *)
+    true_vars : VarSet.t
+  }
+
+type consistent_env =
+  { classes : D.dim UF.t;
+    (** equivalence classes of dimensions from reduced cofibrations *)
 
-    status : [`Consistent | `Inconsistent];
-    (** a cache which must be maintained *)
+    true_vars : VarSet.t;
 
-    true_vars : VarSet.t
+    unreduced_joins : D.cof list list;
+    (** a stack of unreduced joins, each represented by a list of cofibrations *)
   }
 
-let init () =
-  {classes = UF.init ~size:100;
-   true_vars = VarSet.empty;
-   cof = Cof.top;
-   status = `Consistent}
+type env = [ `Consistent of consistent_env | `Inconsistent ]
+
+let init () = `Consistent
+    {classes = UF.init ~size:100;
+     true_vars = VarSet.empty;
+     unreduced_joins = []}
 
-let inconsistent =
-  {classes = UF.init ~size:0;
-   cof = Cof.bot;
-   true_vars = VarSet.empty;
-   status = `Inconsistent}
+let inconsistent = `Inconsistent
 
-let status env = env.status
+let already_inconsistent =
+  function
+  | `Consistent _ -> `Consistent
+  | `Inconsistent -> `Inconsistent
 
 let find_class classes r =
   try UF.find r classes with _ -> r
 
-let unreduced_assumptions env =
-  env.cof
-
-
-module Inversion :
+module type SEQ =
 sig
-  type local =
-    {classes : D.dim UF.t;
-     true_vars : VarSet.t}
+  type t
+  val vacuous : t
+  val seq : ('a -> t) -> 'a list -> t
+end
 
-  val left : local -> D.cof list -> D.cof -> bool
+module rec Search : functor (M : SEQ) ->
```

Comment:
I wonder, can this be written `module rec Search (M : Seq)`? or is it something weird to do with recursive module?

Reply:
It's because the consistency testing is already using the search mechanism.